language/move-prover/src/spec_translator.rs

```diff
@@ -2147,112 +2164,90 @@ impl<'env> SpecTranslator<'env> {
-                let quant_var = self.fresh_var_name("i");
-                let mut is_vector = false;
-                let mut is_domain: Option<Type> = None;
-                match quant_ty {
-                    Type::Vector(..) => is_vector = true,
-                    Type::TypeDomain(t) => is_domain = Some(t.as_ref().clone()),
-                    Type::Primitive(PrimitiveType::Range) => (),
-                    Type::Reference(_, b) => {
-                        if let Type::Vector(..) = *b {
-                            is_vector = true
-                        } else {
-                            panic!("unexpected type")
-                        }
-                    }
-                    _ => panic!("unexpected type"),
-                };
-                if let Some(domain_ty) = is_domain {
-                    let type_check = boogie_well_formed_expr(
-                        self.global_env(),
-                        &var_name,
-                        &domain_ty,
-                        WellFormedMode::WithInvariant,
-                    );
-                    if type_check.is_empty() {
-                        let tctx = TypeDisplayContext::WithEnv {
-                            env: self.global_env(),
-                            type_param_names: None,
-                        };
-                        self.error(
-                            loc,
-                            &format!(
-                                "cannot quantify over `{}` because the type is not concrete",
-                                Type::TypeDomain(Box::new(domain_ty)).display(&tctx)
-                            ),
-                        );
-                    } else {
-                        emit!(
-                            self.writer,
-                            "$Boolean(({} {}: $Value :: {} {} ",
-                            if is_all { "forall" } else { "exists" },
-                            var_name,
-                            type_check,
-                            connective
-                        );
-                        emit!(self.writer, "b#$Boolean(");
-                        self.translate_exp(exp.as_ref());
-                        emit!(self.writer, ")))");
-                    }
+        let quant_ty = self.module_env().env.get_node_type(range.node_id());
+        let connective = match kind {
+            QuantKind::Forall => "==>",
+            QuantKind::Exists => "&&",
+        };
+        let var_name = self.module_env().symbol_pool().string(var.name);
+        let quant_var = self.fresh_var_name("i");
+        let mut is_vector = false;
+        let mut is_domain: Option<Type> = None;
+        match quant_ty {
+            Type::Vector(..) => is_vector = true,
+            Type::TypeDomain(t) => is_domain = Some(t.as_ref().clone()),
+            Type::Primitive(PrimitiveType::Range) => (),
+            Type::Reference(_, b) => {
```

Comment:
nit: you can do `match quant_ty.skip_references()`. In any case, you could also assert on that this is not happening, because the spec language does not allow it. (We may have references from locals coming from the Move lang only.)

Reply:
Thanks, I meant to ask about this actually. The current type checking (before and after PR) doesn't allow references indeed, so maybe I'll just remove this case.

Reply:
Turns out the `.skip_reference()` is useful.